File: crates/harness-server/src/mcp_routes.rs
//! HTTP routes for runtime MCP server management.
//!
//! `GET    /v1/mcp/servers`              — list every registered server
//! `POST   /v1/mcp/servers`              — add a new server
//! `GET    /v1/mcp/servers/:prefix`      — fetch one server's info
//! `DELETE /v1/mcp/servers/:prefix`      — remove + shut down
//! `PUT    /v1/mcp/servers/:prefix`      — replace config in place
//! `POST   /v1/mcp/servers/:prefix/health` — probe with `tools/list`
//!
//! All endpoints require an MCP manager to be configured on
//! `AppState`. When absent, every route returns 503 so callers can
//! cleanly distinguish "not configured" from "really broken".

use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Json, Response},
    routing::{get, post},
    Router,
};
use harness_mcp::{McpClientConfig, McpManager};
use serde_json::json;
use std::sync::Arc;

use crate::state::AppState;
use crate::state_layers::McpLayer;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/v1/mcp/servers", get(list).post(add))
        .route(
            "/v1/mcp/servers/:prefix",
            get(get_one).delete(remove).put(replace),
        )
        .route("/v1/mcp/servers/:prefix/health", post(health))
        .route("/v1/mcp/servers/:prefix/reload", post(reload))
}

#[allow(clippy::result_large_err)]
fn require_mcp(mcp: &McpLayer) -> Result<Arc<McpManager>, Response> {
    mcp.manager.clone().ok_or_else(|| {
        (
            StatusCode::SERVICE_UNAVAILABLE,
            Json(json!({ "error": "mcp manager not configured" })),
        )
            .into_response()
    })
}

async fn list(State(mcp): State<McpLayer>) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    let servers = mcp.list().await;
    (StatusCode::OK, Json(json!({ "servers": servers }))).into_response()
}

async fn get_one(State(mcp): State<McpLayer>, Path(prefix): Path<String>) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    match mcp.get(&prefix).await {
        Some(info) => (StatusCode::OK, Json(json!(info))).into_response(),
        None => (
            StatusCode::NOT_FOUND,
            Json(json!({ "error": "no such mcp server", "prefix": prefix })),
        )
            .into_response(),
    }
}

async fn add(State(mcp): State<McpLayer>, Json(cfg): Json<McpClientConfig>) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    if cfg.prefix.trim().is_empty() {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({ "error": "prefix is required" })),
        )
            .into_response();
    }
    let prefix = cfg.prefix.clone();
    match mcp.add(cfg).await {
        Ok(tools) => (
            StatusCode::CREATED,
            Json(json!({ "prefix": prefix, "tools": tools })),
        )
            .into_response(),
        Err(e) => map_mcp_error(e),
    }
}

async fn replace(
    State(mcp): State<McpLayer>,
    Path(prefix): Path<String>,
    Json(mut cfg): Json<McpClientConfig>,
) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    // Allow callers to omit the prefix in the body — we trust the
    // path. If they DO send one, it must match.
    if cfg.prefix.trim().is_empty() {
        cfg.prefix = prefix.clone();
    } else if cfg.prefix != prefix {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({
                "error": "body prefix does not match path",
                "path": prefix,
                "body": cfg.prefix,
            })),
        )
            .into_response();
    }
    match mcp.replace(&prefix, cfg).await {
        Ok(tools) => (
            StatusCode::OK,
            Json(json!({ "prefix": prefix, "tools": tools })),
        )
            .into_response(),
        Err(e) => map_mcp_error(e),
    }
}

async fn remove(State(mcp): State<McpLayer>, Path(prefix): Path<String>) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    match mcp.remove(&prefix).await {
        Ok(true) => (
            StatusCode::OK,
            Json(json!({ "deleted": true, "prefix": prefix })),
        )
            .into_response(),
        Ok(false) => (
            StatusCode::NOT_FOUND,
            Json(json!({ "error": "no such mcp server", "prefix": prefix })),
        )
            .into_response(),
        Err(e) => map_mcp_error(e),
    }
}

async fn health(State(mcp): State<McpLayer>, Path(prefix): Path<String>) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    let started = std::time::Instant::now();
    match mcp.health(&prefix).await {
        Ok(count) => {
            let latency_ms = started.elapsed().as_millis() as u64;
            (
                StatusCode::OK,
                Json(json!({ "ok": true, "tools": count, "latency_ms": latency_ms })),
            )
                .into_response()
        }
        Err(e) => {
            let latency_ms = started.elapsed().as_millis() as u64;
            (
                StatusCode::OK,
                Json(json!({ "ok": false, "error": e.to_string(), "latency_ms": latency_ms })),
            )
                .into_response()
        }
    }
}

/// `POST /v1/mcp/servers/:prefix/reload` — restart the connection
/// using the slot's currently-stored config. Useful when an MCP
/// server stalls or its child process crashed and the slot fell
/// to `Unhealthy` / `Stopped`. Idempotent: hitting an already-
/// running server is just a quick teardown + reconnect.
async fn reload(State(mcp): State<McpLayer>, Path(prefix): Path<String>) -> Response {
    let mcp = match require_mcp(&mcp) {
        Ok(m) => m,
        Err(r) => return r,
    };
    let started = std::time::Instant::now();
    match mcp.reload(&prefix).await {
        Ok(tools) => {
            let latency_ms = started.elapsed().as_millis() as u64;
            (
                StatusCode::OK,
                Json(json!({
                    "prefix": prefix,
                    "tools": tools,
                    "latency_ms": latency_ms,
                })),
            )
                .into_response()
        }
        Err(e) => map_mcp_error(e),
    }
}

fn map_mcp_error(e: harness_mcp::McpError) -> Response {
    let msg = e.to_string();
    let status = if msg.contains("already registered") || msg.contains("would conflict") {
        StatusCode::CONFLICT
    } else if msg.contains("unknown mcp prefix") || msg.contains("does not match") {
        StatusCode::BAD_REQUEST
    } else {
        StatusCode::INTERNAL_SERVER_ERROR
    };
    (status, Json(json!({ "error": msg }))).into_response()
}

File: packages/server/src/mcp-routes.ts
// REST routes for the MCP server manager. Ported from harness-server's
// mcp_routes.rs. Mounted only when `state.mcpManager` is configured; 503s
// otherwise. The web Settings → McpSection drives all six.
//
//   GET    /v1/mcp/servers                 — { servers: McpServerInfo[] }
//   POST   /v1/mcp/servers                 — add (201 { prefix, tools })
//   GET    /v1/mcp/servers/:prefix         — one McpServerInfo (404 if absent)
//   PUT    /v1/mcp/servers/:prefix         — replace config ({ prefix, tools })
//   DELETE /v1/mcp/servers/:prefix         — remove ({ deleted, prefix } / 404)
//   POST   /v1/mcp/servers/:prefix/health  — probe ({ ok, tools?, latency_ms, error? }, always 200)
//   POST   /v1/mcp/servers/:prefix/reload  — restart ({ prefix, tools, latency_ms })
import type { FastifyInstance, FastifyReply } from "fastify";
import {
  McpManager,
  McpManagerError,
  wireToConfig,
  type McpClientConfigWire,
} from "./mcp-manager.ts";
import type { AppState } from "./state.ts";

function requireManager(state: AppState, reply: FastifyReply): McpManager | undefined {
  if (!state.mcpManager) {
    reply.code(503).send({ error: "mcp manager not configured" });
    return undefined;
  }
  return state.mcpManager;
}

function mapError(reply: FastifyReply, e: unknown): FastifyReply {
  if (e instanceof McpManagerError) return reply.code(e.status).send({ error: e.message });
  return reply.code(500).send({ error: e instanceof Error ? e.message : String(e) });
}

function readConfigBody(reply: FastifyReply, body: unknown): McpClientConfigWire | undefined {
  const b = body as McpClientConfigWire | undefined;
  if (!b || typeof b.prefix !== "string" || typeof b.transport !== "object" || b.transport === null) {
    reply.code(400).send({ error: "body must be a McpClientConfig { prefix, transport, ... }" });
    return undefined;
  }
  return b;
}

export function registerMcpRoutes(app: FastifyInstance, state: AppState): void {
  app.get("/v1/mcp/servers", async (_req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    return reply.send({ servers: m.list() });
  });

  app.post("/v1/mcp/servers", async (req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    const body = readConfigBody(reply, req.body);
    if (!body) return reply;
    try {
      const tools = await m.add(wireToConfig(body));
      return reply.code(201).send({ prefix: body.prefix, tools });
    } catch (e) {
      return mapError(reply, e);
    }
  });

  app.get("/v1/mcp/servers/:prefix", async (req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    const prefix = (req.params as { prefix: string }).prefix;
    const info = m.get(prefix);
    if (!info) return reply.code(404).send({ error: `unknown mcp server '${prefix}'` });
    return reply.send(info);
  });

  app.put("/v1/mcp/servers/:prefix", async (req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    const prefix = (req.params as { prefix: string }).prefix;
    const body = readConfigBody(reply, req.body);
    if (!body) return reply;
    if (body.prefix !== prefix) {
      return reply
        .code(400)
        .send({ error: `body prefix '${body.prefix}' does not match path '${prefix}'` });
    }
    try {
      const tools = await m.replace(prefix, wireToConfig(body));
      return reply.send({ prefix, tools });
    } catch (e) {
      return mapError(reply, e);
    }
  });

  app.delete("/v1/mcp/servers/:prefix", async (req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    const prefix = (req.params as { prefix: string }).prefix;
    const deleted = await m.remove(prefix);
    if (!deleted) return reply.code(404).send({ error: `unknown mcp server '${prefix}'`, prefix });
    return reply.send({ deleted: true, prefix });
  });

  app.post("/v1/mcp/servers/:prefix/health", async (req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    const prefix = (req.params as { prefix: string }).prefix;
    const start = Date.now();
    try {
      const tools = await m.health(prefix);
      return reply.send({ ok: true, tools, latency_ms: Date.now() - start });
    } catch (e) {
      // Health failures are reported in-body with a 200 (the web's McpHealth
      // expects `{ ok: false, error }`), not as an HTTP error.
      return reply.send({
        ok: false,
        error: e instanceof Error ? e.message : String(e),
        latency_ms: Date.now() - start,
      });
    }
  });

  app.post("/v1/mcp/servers/:prefix/reload", async (req, reply) => {
    const m = requireManager(state, reply);
    if (!m) return reply;
    const prefix = (req.params as { prefix: string }).prefix;
    const start = Date.now();
    try {
      const tools = await m.reload(prefix);
      return reply.send({ prefix, tools, latency_ms: Date.now() - start });
    } catch (e) {
      return mapError(reply, e);
    }
  });
}
